### ISSUE TITLE
Automatic sync of downloaded cheatsheets

### ISSUE THREAD

**Issue by papanito** [denisidoro/navi#733]
**Is your feature request related to a problem? Please describe.**

As mentioned at [auto-updating-repositories](https://github.com/denisidoro/navi/blob/master/docs/cheatsheet_repositories.md#auto-updating-repositories), so in case these repos are updated, one can easily get the latest and greatest.

**Describe the solution you'd like**

A command like `navi repo sync` would synchronize i.e. `git pull` the content of the repos

Adding of repos in should be basically a git clone of the cheatsheet repo rather doing a git clone to the `tmp` folder and then copying files over.

**Describe alternatives you've considered**

I currently use and adaption of your [proposal](https://github.com/denisidoro/navi/blob/master/docs/cheatsheet_repositories.md#auto-updating-repositories),

```bash
#!/bin/bash
CHEATS_PATH=$(navi info cheats-path)

cd $CHEATS_PATH

for REPO_PATH in $(ls -d *); do
    echo "Get changes from $REPO_PATH"
    cd $REPO_PATH
    if [ -d ".git" ]; then
        DEFAULT_BRANCH=$(git branch -r | grep -Po 'HEAD -> \K.*$' | cut -d'/' -f2)
        echo $DEFAULT_BRANCH
        git pull -q origin $DEFAULT_BRANCH
    else
        repo_details=($(echo $REPO_PATH | sed  's/__/ /g'))
        repo_user=${repo_details[0]}
        repo_name=$(echo ${repo_details[1]} | cut -d'/' -f1)
        rm *
        git clone "https://github.com/${repo_user}/${repo_name}" .
    fi
    cd ..
done
```

**Additional context**
N/A

**Comment by denisidoro** [comment]
Nice suggestion. It should be easy to implement.

I can't give an ETA, unfortunately.

I'll gladly review a PR, though.

**Comment by alexis-opolka** [comment]
Hi @papanito ,
I'm working on this feature and I would like to have your opinion on what I think would be best.

1. Modify the `navi repo add` subcommand to take the .git folder along with the cheat files instead of cloning the repository as a whole.

Considering that Navi allows you to clone any cheatsheet repository, we should take into account that there might be users who will clone repositories in the wild. This includes the possibility of having malicious code incorporated inside non cheatsheet files.

There is also a pros to it which is that we don't need to separate the logic for someone who wants to select some cheat files and someone who wants all cheat files.

2. Add a `navi repo sync` subcommand which will have two behaviours.

The first will be when called without arguments, it will iterate over all added cheatsheet repositories and git pull the latest changes.

The second will be when called with an argument (e.g. the name of the repository), it will search the cheatsheet repository and if found git pull the latest changes.

N.B. It might be interesting to add a `navi repo list` to list the installed repositories instead of making the user do a ls on the cheats path.

**Comment by papanito** [comment]
Thanks @alexis-opolka glad you value my input ;-)

I actually like the idea of 1. and 2. sounds reasonable and makes sense imho. @denisidoro?

> Considering that Navi allows you to clone any cheatsheet repository, we should take into account that there might be users who will clone repositories in the wild. This includes the possibility of having malicious code incorporated inside non cheatsheet files.

This is basically true wherever a user clones a repo... Question is how to ensure that only cheat files are cloned? A simple warning before adding or syncing a repo, might be a start.

**Comment by alexis-opolka** [comment]
You're the author of this feature request, I find it reasonable to ask you what you think about how it's going to be implemented. ^-^

Regarding the security issue, we're currently only importing files with `.cheat` or `.cheat.md` extensions and discarding the rest.

What we can do is to "streamline" the import process:

- get rid of the yes/no question as-is and transform the screen to be a confirmation prompt like the one package managers can show (e.g. apt, DNF, etc.)

    It will be on this screen that we can show a cautionary notice about using untrusted repositories.  
    It would be nice to be able to store trusted sources so we don't ask each time.

- Make the screen mandatory to select files (which already includes a shortcut to select all)
- Introduce an option, as requested by another issue, to bypass confirmation i.e. `-y` or `-f`
- Introduce another option to select cheat files without interactions.

**Comment by alexis-opolka** [comment]
If my memory serves me right, there is already an issue about the security issues users might encounter with Navi so I will postpone the work on that for now.

Since this part of Navi still needs to be refined and thought of throughout the whole codebase I think more discussions are required before any work can be done.

I can make test builds in the mean time to test features regarding the security concerns with Navi.

**Comment by alexis-opolka** [comment]
Hi @papanito ,
If you'd like to give me feedback regarding the `repo list` and `repo sync` subcommands either regarding the usage or the implementation itself, the code is available on the experimental branch: https://github.com/denisidoro/navi/tree/experimental.

The features should be working correctly, I haven't yet tackled the issue of sorting files when doing a pull so you might have duplicates when doing so. I'll work on that part for the next couple of hours.

**Comment by papanito** [comment]
>  If you'd like a binary pre-release, let me know.

That would be nice. Still would need some time to test it... ;-)

**Comment by alexis-opolka** [comment]
Hi, a quick update on this issue.

I've been unable to work on it seriously these past few weeks to wrap up the last missing bits for the feature to be somewhat stable.
Since my weekend starts tomorrow and lasts for four days (thanks french public holidays XD) I will wrap it up in the following days.

I mainly need to have a stable way of checking if I need to pull changes or not (i.e. having diverged from the remote branch) in order to reduce the number of useless transactions with git and solve some other minor issues.
I already have some ways that would do it but as I said previously I couldn't work on it seriously to know if it works correctly.

I also pushed some other features on the experimental branch that I need to either link to issues and/or document so the next pre-release will be interesting to test. ^-^

**Comment by alexis-opolka** [comment]
Hi @papanito !

Sorry I haven't been able to be really active these last few months but I still worked on it.
Expect the version to be pushed tomorrow or next week at the latest.

**Comment by alexis-opolka** [comment]
I got some false hopes along the way, but I've come to find quite a good way to easily sync the cheatsheets from a git repository.
I'm going to commit and push everything on the experimental branch and create a PR in the following days.

Again, sorry for the long wait. 😅
